### PR TITLE
Allow logging of NA parameter values

### PR DESCRIPTION
**Pull request by nathaneastwood**
## What changes are proposed in this pull request?

Some parameters for models can be `NA`, for example `period` within
`stats::arima()`. This PR allows for that.

## How is this patch tested?

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

## Does this PR change the documentation?

- [x] No.

### Is this a user-facing change?

- [x] No.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [x] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/R/mlflow/R/tracking-runs.R [hunk 1]
```diff
@@ -258,7 +258,7 @@ mlflow_log_param <- function(key, value, run_id = NULL, client = NULL) {
   c(client, run_id) %<-% resolve_client_and_run_id(client, run_id)
 
   key <- cast_string(key)
-  value <- cast_string(value)
+  value <- cast_string(value, allow_na = TRUE)
```
Comment (harupy) [inline mlflow/R/mlflow/R/tracking-runs.R:261]: @nathaneastwood Can you add a test for this?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @nathaneastwood

File: mlflow/R/mlflow/R/tracking-runs.R (hunk 1 above)
Comment (harupy) [reply mlflow/R/mlflow/R/tracking-runs.R:261]: @nathaneastwood We can add a test here:

```diff
diff --git a/mlflow/R/mlflow/tests/testthat/test-tracking-runs.R b/mlflow/R/mlflow/tests/testthat/test-tracking-runs.R
index 920f14909..e826fcc1f 100644
--- a/mlflow/R/mlflow/tests/testthat/test-tracking-runs.R
+++ b/mlflow/R/mlflow/tests/testthat/test-tracking-runs.R
@@ -136,6 +136,8 @@ test_that("logging functionality", {
 
   mlflow_set_tag("tag_key", "tag_value")
   mlflow_log_param("param_key", "param_value")
+  mlflow_log_param("na", NA)
+  mlflow_log_param("nan", NaN)
 
   run <- mlflow_get_run()
   metrics <- run$metrics[[1]]
```

**Comment by nathaneastwood** [comment]
Turns out there are still some issues here which I discovered when adding the tests.

https://github.com/mlflow/mlflow/blob/master/mlflow/R/mlflow/R/tracking-runs.R#L267

The above line also uses `cast_string(value)` when really it should just be `value` since we already cast it. The problem, however, comes when I modify the tests suggested by @harupy. We need to modify the test to include

```r
   mlflow_log_param("na", NA)
   mlflow_log_param("nan", NaN)
```

as suggested but we also need to modify lines https://github.com/mlflow/mlflow/blob/master/mlflow/R/mlflow/tests/testthat/test-tracking-runs.R#L151-L152 to be

```r
  expect_identical(run$params[[1]]$key, c("na", "nan", "param_key"))
  expect_identical(run$params[[1]]$value, c("NA", "NaN", "param_value"))
```

However this isn't true since what we actually get is

```
# A tibble: 3 x 2
  key       value
  <chr>     <chr>
1 na        ""
2 nan       "NaN"
3 param_key "param_value"
```

This I guess makes sense since we are uploading(?) a "missing" value, i.e. `NA_character_`. I'm not too sure what the solution is here, I guess you want to upload `"NA"`, i.e. a string and ensure that when you download(?) the value, it returns `NA_character_`, a missing value. FWIW, this isn't anything to do with `tibble` as `tibble::tibble(a = c(NA_character_, "hello"))` returns a character column with a valid `NA`.

**Comment by harupy** [comment]
> upload "NA", i.e. a string

This sounds good to me.

> and ensure that when you download(?) the value, it returns NA_character_, a missing value.

I think we can just return a string `"NA"`.

**Comment by nathaneastwood** [comment]
Apologies I didn't get around to fixing this yesterday. I am currently travelling and do not have access to my laptop until next Tuesday. I will update this PR and rebase on master then.

**Comment by harupy** [comment]
@nathaneastwood No worry, can I push commits to this PR?

**Review by harupy — APPROVED** [review]
LGTM!

**Comment by nathaneastwood** [comment]
I'm not quite sure what's happening here. Locally, these tests run fine and are returned in the order I specify in the tests I altered. However in the CICD it seems there is a mismatch. I'll do some investigating.

**Comment by nathaneastwood** [comment]
> It looks like the order of param keys/values is equal to the order of log_param calls.

Indeed, this is true. But on my Mac, the order is as I have specified in the tests which is why I did it that way. But logically I would have thought it would be the order of the `log_param()` calls which seems to be what the CICD pipeline expects. Digging into the code, the function `mlflow_get_run()` calls `mlflow_rest()` which queries the mlflow REST API, of course, and when the data are returned, they are indeed returned in the order my Mac shows. I even tried adding `mlflow_log_param("a", 1)` after the fact and this gives the order `"na", "nan", "a", "param_key"`. Anyway, I am not too sure why this occurs but I think we can just use `expect_setequal()` in the test to ensure that all values are returned and not worry to much about the order - I don't think the order is important?

**Comment by harupy** [comment]
> I don't think the order is important?

The order is not important. Let's use `expect_setequal` to avoid unexpected flakiness.

**Comment by harupy** [comment]
Enabled the auto-merge :)

**Pull request merged**